FreeRTOS/FreeRTOS
# FreeRTOS+TCP : add memory statistics and dump packets

See also an earlier equivalent [PR 1828](https://github.com/aws/amazon-freertos/pull/1828) on aws/amazon-freertos

Description
-----------
Sometimes it is difficult to get an overview of how much RAM is actually used by FreeRTOS+TCP.

When enabled, `tcp_mem_stats.c` will write CSV records to the standard logging. With the use of grep and sed, these lines can be taken out and copied into a spreadsheet.

The result is a sheet (with formulas), that shows all relevant ipconfig macros, their values and how many objects are declared, either statically or dynamically. The user can test several scenarios by changing the macro values.

When not enabled ( `ipconfigUSE_TCP_MEM_STATS` not defined ), the module is not compiled and no extra code is added.

I tested the module on several platforms, notably ESP32, Xilinx and WinSim.

There is also some documentation in tools/tcp_mem_stats.md

Also: added `tcp_dump_packets.c` : it can be included in the WinSim project. It will write selected network packets to a C source files. This source file can later be include in testing projects to test the TCP software.

Test Steps
-----------
`tcp_mem_stats.c` : I had it run for a few minutes in a aws_tests project. After `vTCPMemStatClose()` was called, I did the following on the logging:

`tcp_dump_packets.c` : I filled in the desired packet types and had it run until all packet  types were collected. I compiled the output into a testing module.

~~~
grep ".*TCPMemStat," < output.log | sed "s/.*TCPMemStat,//" > output.csv
~~~

I opened the CSV file as a spreadsheet.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
